### DISCUSSION TITLE
How do I read the RGB values of an image?

### DISCUSSION THREAD

**Discussion by copypasteonly** [libvips/libvips#2715] [Q&A]
I'm trying to follow this [guide](https://robertheaton.com/2018/06/12/programming-projects-for-advanced-beginners-ascii-art/) on how to make an ASCII Art program, I'm currently stuck at finding out how I would get the RGB values for every pixel in an image. Am I right to assume that it would have to do something about bands? Since I've read that bands usually represent color. How would I be able to access this? Thanks!

**Comment by jcupitt** [comment]
Hi @copypasteonly,

What language are you using? If it's python, you can just use `image(x, y)`:

```python
import pyvips

x = pyvips.Image.new_from_file("something.jpg")
# jpegs can be eg. CMYK, so make sure we have RGB
x = x.colourspace("srgb")
# read out rgb
r, g, b = x(10, 10)
```

`image(x, y)` is quite slow -- you might want something better for large images.

**Reply by copypasteonly** [reply]
Hello, I'm using C, I'm guessing the syntax would be something like 

```C
#include <stdio.h>
#include <vips/vips.h>

int main( int argc, char **argv ){
    VipsImage *x = vips_image_new_from_file("test", NULL);
    x = vips_colourspace("srgb");
    r, g, b = x(10,10);
}
```
Is that correct? I would also like to ask if how would I be able to access every individual pixel in an image? Would it be something like
`x[0][0][0]` Since it's a 3D Array? If so then what will be the use of the 3rd index, since it's just the "band" of the pixel. Is there anyway I can access the individual r, g, b values through the array? Or do I just use the method given above?

**Reply by jcupitt** [reply]
For C you'd need something like:

```C
/* Compile with:
 *
 *      gcc -g -Wall getpoint.c `pkg-config vips --cflags --libs`
 */

#include <vips/vips.h>

int
main( int argc, char **argv )
{
        GRand *rand;
        VipsImage *image;
        int i;

        if( VIPS_INIT( argv[0] ) )
                vips_error_exit( NULL );

        vips_leak_set( TRUE );

        if( argc == 1 )
                vips_error_exit( "usage:\n\t%s test-image", argv[0] );
        if( !(image = vips_image_new_from_file( argv[1], NULL )) )
                vips_error_exit( NULL );
        if( image->Bands != 3 )
                vips_error_exit( "not a three-band image" );

        rand = g_rand_new();

        for( i = 0; i < 1000; i++ ) {
                int x = g_rand_int_range( rand, 0, image->Xsize );
                int y = g_rand_int_range( rand, 0, image->Ysize );

                double *vector;
                int n;

                if( vips_getpoint( image, &vector, &n, x, y, NULL ) )
                        vips_error_exit( NULL );

                printf( "(%d, %d): %g %g %g\n",
                        x, y, vector[0], vector[1], vector[2] );

                g_free( vector );
        }

        g_rand_free( rand );

        g_object_unref( image );

        vips_shutdown();

        return( 0 );
}
```

Then you can run:

```
$ ./a.out ~/pics/k2.jpg
(353, 1447): 33 35 21
(482, 173): 48 49 69
(1144, 710): 191 142 112
(179, 1996): 106 91 50
...
```

`vips_getpoint()` is documented here:

https://www.libvips.org/API/current/libvips-arithmetic.html#vips-getpoint

**Reply by copypasteonly** [reply]
Thank you!

**Reply by husqD** [reply]
> For C you'd need something like:
> 
> ```c
> /* Compile with:
>  *
>  *      gcc -g -Wall getpoint.c `pkg-config vips --cflags --libs`
>  */
> 
> #include <vips/vips.h>
> 
> int
> main( int argc, char **argv )
> {
>         GRand *rand;
>         VipsImage *image;
>         int i;
> 
>         if( VIPS_INIT( argv[0] ) )
>                 vips_error_exit( NULL );
> 
>         vips_leak_set( TRUE );
> 
>         if( argc == 1 )
>                 vips_error_exit( "usage:\n\t%s test-image", argv[0] );
>         if( !(image = vips_image_new_from_file( argv[1], NULL )) )
>                 vips_error_exit( NULL );
>         if( image->Bands != 3 )
>                 vips_error_exit( "not a three-band image" );
> 
>         rand = g_rand_new();
> 
>         for( i = 0; i < 1000; i++ ) {
>                 int x = g_rand_int_range( rand, 0, image->Xsize );
>                 int y = g_rand_int_range( rand, 0, image->Ysize );
> 
>                 double *vector;
>                 int n;
> 
>                 if( vips_getpoint( image, &vector, &n, x, y, NULL ) )
>                         vips_error_exit( NULL );
> 
>                 printf( "(%d, %d): %g %g %g\n",
>                         x, y, vector[0], vector[1], vector[2] );
> 
>                 g_free( vector );
>         }
> 
>         g_rand_free( rand );
> 
>         g_object_unref( image );
> 
>         vips_shutdown();
> 
>         return( 0 );
> }
> ```
> 
> Then you can run:
> 
> ```
> $ ./a.out ~/pics/k2.jpg
> (353, 1447): 33 35 21
> (482, 173): 48 49 69
> (1144, 710): 191 142 112
> (179, 1996): 106 91 50
> ...
> ```
> 
> `vips_getpoint()` is documented here:
> 
> https://www.libvips.org/API/current/libvips-arithmetic.html#vips-getpoint

Hi @jcupitt, I need to read all the pixels, one by one, but I find getpoint too slow (I use the lib in C++). 
Can you give me some alternative to make the process much faster? I have tried everything possible but I can't find a solution to this problem.

Thanks in advance.

**Reply by jcupitt** [reply]
The fastest way is to write a new libvips operator. Here's `avg`, for example:

https://github.com/libvips/libvips/blob/master/libvips/arithmetic/avg.c

That loops over every pixel in an image computing the average value.

It subclasses this thing, which is what implements the loop using `vips_sink()`:

https://github.com/libvips/libvips/blob/master/libvips/arithmetic/statistic.c

Have a look at the `build` method.

There's a chapter in the docs about this:

https://www.libvips.org/API/current/extending.html

**Reply by husqD** [reply]
> La forma más rápida es escribir un nuevo operador libvips. Aquí está `avg`, por ejemplo:
> 
> https://github.com/libvips/libvips/blob/master/libvips/arithmetic/avg.c
> 
> Eso se repite sobre cada píxel en una imagen calculando el valor promedio.
> 
> Subclasifica esta cosa, que es lo que implementa el bucle usando `vips_sink()`:
> 
> https://github.com/libvips/libvips/blob/master/libvips/arithmetic/statistic.c
> 
> Echa un vistazo al `build` método.
> 
> Hay un capítulo en los documentos sobre esto:
> 
> https://www.libvips.org/API/current/extending.html

Unfortunately at the moment I don't think I will be able to understand the library enough to do what you say, but thank you for taking the time to answer me.
